tailnet/conn.go

```diff
@@ -519,22 +536,35 @@ func (c *Conn) Close() error {
 	default:
 	}
 	close(c.closed)
-	for _, l := range c.listeners {
-		_ = l.closeNoLock()
-	}
 	c.mutex.Unlock()
-	c.dialCancel()
-	_ = c.dialer.Close()
-	_ = c.magicConn.Close()
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	if c.trafficStats != nil {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			_ = c.trafficStats.Shutdown(ctx)
+		}()
+	}
+
 	_ = c.netStack.Close()
+	c.dialCancel()
 	_ = c.wireguardMonitor.Close()
-	_ = c.tunDevice.Close()
+	_ = c.dialer.Close()
+	// Stops internals, e.g. tunDevice, magicConn and dnsManager.
 	c.wireguardEngine.Close()
-	if c.trafficStats != nil {
-		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-		defer cancel()
-		_ = c.trafficStats.Shutdown(ctx)
+
+	c.mutex.Lock()
+	for _, l := range c.listeners {
+		_ = l.closeNoLock()
 	}
+	c.listeners = nil
+	c.mutex.Unlock()
```

Comment:
I doubt these changes are ready, I simply tried to mimic the order a bit closer to what's done in: https://github.com/tailscale/tailscale/blob/cd18bb68a49608b86f60e22cb081f0156d3d11b5/tsnet/tsnet.go#L152-L192